src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -188,5 +255,155 @@ private bool AssemblyContainsTagHelpers(string assemblyName, IReadOnlyList<TagHe
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
+                        {
+                            tagHelperPrefix.Diagnostics.Add(
+                                ComponentDiagnosticFactory.Create_UnsupportedTagHelperDirective(node.GetSourceSpan(_source)));
+                        }
+                    }
+                    else if (context.ChunkGenerator is AddImportChunkGenerator usingStatement && !usingStatement.IsStatic)
+                    {
+                        // Get the namespace from the using statement. Split it at '=' in case there is an alias.
+                        var @namespace = usingStatement.ParsedNamespace.Split('=').LastOrDefault();
+                        for (var i = 0; i < _tagHelpers.Count; i++)
+                        {
+                            var tagHelper = _tagHelpers[i];
+                            if (tagHelper.IsComponentFullyQualifiedNameMatch())
+                            {
+                                // We've already added these to our list of matches.
+                                continue;
+                            }
+
+                            var typeName = tagHelper.GetTypeName();
+                            if (typeName != null && IsTypeInNamespace(typeName, @namespace))
+                            {
+                                // If the type is at the top-level or if the type's namespace matches the using's namespace, add it.
+                                Matches.Add(tagHelper);
+                            }
+                        }
+                    }
+                }
+            }
+
+            private bool IsTypeInNamespace(string typeName, string @namespace)
+            {
+                var namespaceLength = typeName.LastIndexOf('.');
+                if (namespaceLength == -1)
+                {
+                    // Either the typeName is not the full type name or this type is at the top level.
+                    return true;
+                }
+
+                var typeNamespace = typeName.Substring(0, namespaceLength);
+                if (typeNamespace.Equals(@namespace, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool IsTypeInScope(string typeName, string currentNamespace)
+            {
+                var namespaceLength = typeName.LastIndexOf('.');
+                if (namespaceLength == -1)
+                {
+                    // Either the typeName is not the full type name or this type is at the top level.
+                    return true;
+                }
+
+                var typeNamespace = typeName.Substring(0, namespaceLength);
+                var typeNamespaceSegments = typeNamespace.Split(NamespaceSeparators, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
Shouldn't remove empty entries on these because if the user typed a bad namespace (adjacent dots) it could unintentionally match.

Reply:
Actually the user doesn't type either of these namespaces. One comes from RootNamespace+filepath and the other one comes from the tag helper descriptor. I think it's fine to be extra defensive here.